Source/Core/InputCommon/ControllerInterface/GameController/GameController.mm

```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <random>
+
+#import <objc/message.h>
+
+#include "InputCommon/ControllerInterface/GameController/GameController.h"
+#include "InputCommon/ControllerInterface/GameController/GameControllerDevice.h"
+
+#include "Common/Logging/Log.h"
+
+namespace ciface::GameController
+{
+static char *s_observer="";
+
+void addController(GCController* controller)
+{
+  controller.motion.sensorsActive = true;
+
+  std::string vendorName = std::string([controller.vendorName UTF8String]);
+  INFO_LOG_FMT(SERIALINTERFACE, "Controller '{}' connected with motion={}", vendorName, controller.motion.sensorsActive);
+  NOTICE_LOG_FMT(SERIALINTERFACE, "Controller '{}' connected with motion={} and id=", vendorName, controller.motion.sensorsActive);
+
+  std::shared_ptr<ciface::GameController::Controller> j=std::make_shared<Controller>(controller);
+    
+  g_controller_interface.AddDevice(j);
+}
+
+static void onControllerConnect(CFNotificationCenterRef center, void* observer, CFStringRef name, const void* object, CFDictionaryRef userInfo)
+{
+  GCController* controller = (GCController*)object;
+  addController(controller);
+}
+
+static void onControllerDisconnect(CFNotificationCenterRef center, void* observer, CFStringRef name, const void* object, CFDictionaryRef userInfo)
+{
+  GCController *inController = (GCController*)object;
+    
+  std::string vendorName = std::string([inController.vendorName UTF8String]);
+  INFO_LOG_FMT(SERIALINTERFACE, "Controller '{}' disconnected", vendorName);
+    
+  g_controller_interface.RemoveDevice([&inController](const auto* obj)
+  {
+    const Controller* controller = dynamic_cast<const Controller*>(obj);
+    if (controller && controller->IsSameDevice(inController))
+      return true;
+
+    return false;
+  });
+}
+
+void Init(void* window)
+{
+  CFNotificationCenterAddObserver(CFNotificationCenterGetLocalCenter(), s_observer, &onControllerConnect, CFSTR("GCControllerDidConnectNotification"), NULL, CFNotificationSuspensionBehaviorDeliverImmediately);
```

Comment:
Prefer using `nullptr` over NULL. Ditto for other applicable cases elsewhere.

Reply:
Does this trigger the callback for devices that are connection on Init, or does it only call it for devices that will be connected after binding to it? I guess it's the second case, but please specify it in a comment.

Reply:
Also there is a space between the two functions here, but there isn't in DeInit()